quicktill/pdrivers.py

```diff
@@ -701,6 +716,38 @@ def process_canvas(self, canvas, f):
                 f.write(escpos.ep_fullcut)
             f.flush()
 
+    def _image(self, data, width, height, f):
+        # Print a PBM bit-image
+        if width > self.dpl:
+            # Image too wide for paper
+            return
+
+        # Partition the bitmap into rows
+        rows = []
+        for line in range(height):
+            start = line * (width // 8)
+            end = start + (width // 8)
+            rows.append(data[start:end])
+
+        # Calculate padding required to center the image
+        padding = (self.dpl - width) // 2
+        padchars = bytes([False]) * padding
+
+        # Write the commands to render the padded image
+        f.write(escpos.ep_unidirectional_on)
+        for row in rows:
+            bits = []
+            for byte in row:
+                for bit in f"{int(byte):08b}":
+                    bits.extend([bool(int(bit))] * 3)
+            header = escpos.ep_bitimage_sd + bytes([len(bits), 1])
```

Comment:
Here I get `ValueError: bytes must be in range(0, 256)`

Reply:
My test image is 512×512, remember.

Reply:
Ok, thanks.  I guess there are two things to understand here:

* Is there an upper-bound that we want to place on the image sizes allowed for logos?
* Does the ESC/P format allow for widths/heights greater than `256`?  If so, that could open various other questions (what endianness do we write the width/height dimensions in, and how do we indicate to the printer how many bytes to read when loading each dimension)

Reply:
`ep_bitimage_sd` is followed by 16 bits of image width data, so we can go to width 65536 if necessary! And to go taller we just send multiple rows.

Reply:
Oddly the documentation indicates `nL + nH * 256` (addition, not multiplication of the two bytes) - isn't that up to 131072?

Reply:
Hrm, in fact usually the multiplication would have higher precedence; my mistake.

Reply:
512x512 images are now read and printed, _provided that they do not exceed the dots-per-line value for the printer_ (currently nothing is printed as logo output when that happens).

I'm still at a bit of a loss about how to get the output to print in a reasonable format though.

Reply:
Perhaps I should begin using `pillow` to scale the image down to a reasonable size for centre-aligned display?

(I don't think that'd fix the scanline-like appearance of the printout, but it would solve/bypass the width-limit issue)